File: rust/src/redis/group_make.rs
use ::redis::AsyncTypedCommands;
use ::redis::RedisResult;

/// Creates a consumer group for a Redis stream, ensuring idempotent behavior.
///
/// If the group already exists, the function returns `Ok(())` instead of failing.
/// This is achieved by ignoring "BUSYGROUP" errors, which are returned when attempting
/// to create a group that already exists.
///
/// # Arguments
///
/// * `con` - An async-capable Redis command connection
/// * `stream_name` - The name of the Redis stream
/// * `group_name` - The name of the consumer group to create
///
/// # Returns
///
/// * `Ok(())` - If the group was created or already exists
/// * `Err(RedisError)` - If the operation fails for any reason other than group already existing
pub(super) async fn make_stream_group(
  mut con: impl AsyncTypedCommands,
  stream_name: impl Into<String>,
  group_name: impl Into<String>,
) -> RedisResult<()> {
  match con
    .xgroup_create_mkstream(stream_name.into(), group_name.into(), "$")
    .await
  {
    Ok(_) => Ok(()),
    Err(err) => {
      // Ignore "BUSYGROUP" errors (group already exists) to make subscription idempotent.
      if err.code() == Some("BUSYGROUP") {
        Ok(())
      } else {
        Err(err)
      }
    }
  }
}

File: go/redis/group_make.go
package redis

import (
	"context"
	"strings"

	"github.com/redis/go-redis/v9"
)

// MakeStreamGroup creates a consumer group for a Redis stream, ensuring idempotent behavior.
//
// If the group already exists, the function returns nil instead of failing.
// This is achieved by ignoring "BUSYGROUP" errors, which are returned when attempting
// to create a group that already exists.
//
// # Arguments
//
//   - ctx: A context for cancellation and timeouts
//   - cmd: A Redis command interface (typically *redis.Client or redis.Cmdable)
//   - streamName: The name of the Redis stream
//   - groupName: The name of the consumer group to create
//
// # Returns
//
//   - nil - If the group was created or already exists
//   - error - If the operation fails for any reason other than group already existing
func MakeStreamGroup(
	ctx context.Context,
	cmd redis.Cmdable,
	streamName, groupName string,
) error {
	err := cmd.XGroupCreateMkStream(ctx, streamName, groupName, "$").Err()
	if err != nil {
		// Ignore "BUSYGROUP" errors (group already exists) to make subscription idempotent.
		if strings.Contains(err.Error(), "BUSYGROUP") {
			return nil
		}
		return err
	}
	return nil
}
